bench/bench.ml

```diff
@@ -0,0 +1,224 @@
+
+  let variance xs =
+    let xs_mean = mean xs in
+    List.map (fun x -> Float.pow (x -. xs_mean) 2.) xs |> mean
+
+  let stddev xs = Float.sqrt (variance xs)
+end
+
+module Benchmark = struct
+  type t = { driver : Driver.t; input : Input.t }
+
+  let create ~driver ~input = { driver; input }
+
+  let name { driver; input } =
+    Printf.sprintf "%s %s" (Driver.name driver) (Input.name input)
+
+  let time_run_blocking { driver; input } =
+    time_run_blocking driver.path [ input.path ]
+
+  let repeat n f = List.init n (fun _ -> f ())
+
+  let run_blocking t ~n_warmup ~n =
+    let run () = time_run_blocking t in
+    let _warmup = repeat n_warmup run in
+    let times = repeat n run in
+    let mean = Stats.mean times in
+    let stddev = Stats.stddev times in
+    let metrics =
+      [
+        Output.Metric.create ~name:"time mean" ~value:(`Float mean)
+          ~units:"seconds";
+        Output.Metric.create ~name:"time stddev" ~value:(`Float stddev)
+          ~units:"seconds";
+      ]
+    in
+    Output.Result.create ~name:(name t) ~metrics
+end
+
+module Driver_dir = struct
+  type t = { path : string }
+
+  let driver_name = "driver.exe"
+  let inputs_dir_name = "inputs"
+  let driver_path { path } = Filename.concat path driver_name
+  let driver t = { Driver.path = driver_path t }
+  let inputs_path { path } = Filename.concat path inputs_dir_name
+
+  let of_path path =
+    let t = { path } in
+    if not (Sys.file_exists (driver_path t)) then
+      failwith (Printf.sprintf "failed to find %s in %s" driver_name path);
+    if not (Sys.file_exists (inputs_path t)) then
+      failwith (Printf.sprintf "failed to find %s in %s" inputs_dir_name path);
+    t
+
+  let inputs t =
+    readdir_full_paths (inputs_path t) |> List.map (fun path -> { Input.path })
+
+  let benchmarks t =
+    let driver = driver t in
+    inputs t |> List.map (fun input -> Benchmark.create ~driver ~input)
+end
+
+module Path_to_current_exe = struct
+  let via_procfs () =
+    (* Look up the current executable's path in the proc filesystem. *)
+    let pid = Unix.getpid () in
+    let proc_exe_path = Printf.sprintf "/proc/%d/exe" pid in
+    if Sys.file_exists proc_exe_path then Some (Unix.readlink proc_exe_path)
+    else None
+
+  let via_cwd () =
+    (* Assume the current working directory is the root of the project (as it
+       would be if this was run via `make bench`) and find the path to the
+       current exe relative to the project root *)
+    let cwd = Unix.getcwd () in
+    let relative_path_parts = "_build/default/bench/bench.exe" in
+    let absolute_path = Filename.concat cwd relative_path_parts in
+    if Sys.file_exists absolute_path then Some absolute_path else None
+
+  let methods = [ via_procfs; via_cwd ]
+
+  let get () =
+    let maybe_path = List.find_map (fun m -> m ()) methods in
+    match maybe_path with
+    | Some path -> path
+    | None -> failwith "couldn't determine the path to the current exe"
+end
```

Comment:
This code is just to deal with the case where bench is called somewhere else than the root of the project. That's a nice attention to details :+1:

Reply:
Thanks!